test/unit_tests/braket/experimental/autoqasm/test_parameters.py

```diff
@@ -289,3 +289,142 @@ def my_program():
+    assert bound_prog.to_ir() == bound_template.format(0.432143)
+
+
+def test_multi_bind_parameters():
+    """Test binding FreeParameters to concrete values."""
+
+    @aq.subroutine
+    def sub(alpha: float, theta: float):
+        rx(0, alpha)
+        rx(1, theta)
+        cnot(0, 1)
+        rx(0, theta)
+        rx(1, alpha)
+
+    @aq.subroutine
+    def rx_alpha(qubit: int):
+        rx(qubit, FreeParameter("alpha"))
+
+    @aq.main(num_qubits=3)
+    def parametric(alpha: float, theta: float):
+        sub(alpha, theta)
+        rx_alpha(2)
+
+    prog = parametric(FreeParameter("alpha"), FreeParameter("beta"))
+    bound_prog = prog.make_bound_program({"alpha": 0.5, "beta": 1.5})
+
+    expected = """OPENQASM 3.0;
+def sub(float[64] alpha, float[64] theta) {
+    rx(alpha) __qubits__[0];
+    rx(theta) __qubits__[1];
+    cnot __qubits__[0], __qubits__[1];
+    rx(theta) __qubits__[0];
+    rx(alpha) __qubits__[1];
+}
+def rx_alpha(int[32] qubit) {
+    rx(alpha) __qubits__[qubit];
+}
+float[64] alpha = 0.5;
+float[64] beta = 1.5;
+qubit[3] __qubits__;
+sub(alpha, beta);
+rx_alpha(2);"""
+    assert bound_prog.to_ir() == expected
+
+
+def test_partial_bind():
+    """Test binding some but not all FreeParameters."""
+
+    @aq.subroutine
+    def rx_alpha(qubit: int, theta: float):
+        rx(qubit, theta)
+
+    @aq.main(num_qubits=3)
+    def parametric(alpha: float, beta: float):
+        rx_alpha(2, alpha)
+        rx_alpha(2, beta)
+
+    prog = parametric(FreeParameter("alpha"), FreeParameter("beta"))
+    bound_prog = prog.make_bound_program({"beta": np.pi})
+
+    expected = """OPENQASM 3.0;
+def rx_alpha(int[32] qubit, float[64] theta) {
+    rx(theta) __qubits__[qubit];
+}
+input float[64] alpha;
+float[64] beta = pi;
+qubit[3] __qubits__;
+rx_alpha(2, alpha);
+rx_alpha(2, beta);"""
+    assert bound_prog.to_ir() == expected
+
+
+def test_binding_pulse_parameters():
+    """Test binding programs with parametric pulse instructions."""
+
+    @aq.gate_calibration(implements=rx, target="$1")
+    def cal_1(angle: float):
+        pulse.delay("$1", angle)
+
+    @aq.main
+    def my_program():
+        rx("$1", FreeParameter("theta"))
+
+    qasm1 = my_program().with_calibrations(cal_1).make_bound_program({"theta": 0.6}).to_ir()
+    qasm2 = my_program().make_bound_program({"theta": 0.6}).with_calibrations(cal_1).to_ir()
+    assert qasm1 == qasm2
+
+    expected = """OPENQASM 3.0;
+defcal rx(angle[32] angle) $1 {
+    delay[(angle) * 1s] $1;
+}
+float[64] theta = 0.6;
+rx(theta) $1;"""
+    assert expected == qasm1
+
+
+def test_bind_empty_program():
+    """Test that binding behaves well on empty programs."""
+
+    @aq.main
+    def empty_program():
+        pass
+
+    qasm = empty_program().to_ir()
+    bound_program1 = empty_program().make_bound_program({}).to_ir()
+    bound_program2 = empty_program().make_bound_program({"alpha": 0.5}).to_ir()
+    assert qasm == bound_program1 == bound_program2
```

Comment:
cool, I had somehow never realized that Python allowed chaining equality tests like this.

Reply:
I wasn't sure either :D I followed a whim